Support (s,?,o) queries for KvsAll training

vonVogelstein opened the pull request.

vonVogelstein commented:
As discussed with Rainer, added the option "use_os" for KvsAll training. The option allows to train not only on the rows and columns created from the sp and po pairs, but also on the tubes created from the so pairs

rgemulla commented on kge/config-default.yaml:
```diff
@@ -220,6 +220,9 @@ KvsAll:
   # ( https://github.com/TimDettmers/ConvE/blob/853d7010a4b569c3d24c9e6eeaf9dc3535d78832/main.py#L156) with a default value of 0.1.
   label_smoothing: 0.0
 
+  # If set, training also runs on subject object (so) pairs additionally to sp and op pairs
```
Please rebase to master and then change options to: 
```
queries:
   sp_: True
   s_o: False
   _po: True
```
I.e., each of the KvsAll prediction tasks can be enabled/disabled individually.

rgemulla commented on kge/job/train.py:
```diff
@@ -583,55 +596,65 @@ def collate(batch):
                     values = self.train_sp_values
-                    sp_po_col_1, sp_po_col_2, o_s_col = S, P, O
-                else:
+                    sp_po_so_col_1, sp_po_so_col_2, o_s_p_col = S, P, O
```
try to use more intuitive variable names: query_col1, query_col2, target_col for example

rgemulla commented on kge/job/train.py:
```diff
@@ -583,55 +596,65 @@ def collate(batch):
+                "sp_po_so_batch": sp_po_so_batch,
                 "label_coords": label_coords,
-                "is_sp": is_sp,
+                "key_type": key_type,
```
query_type

rgemulla commented on kge/job/train.py:
```diff
@@ -672,15 +706,28 @@ def _process_batch(self, batch_index, batch) -> TrainingJob._ProcessBatchResult:
             loss_value_po.backward()
             backward_time += time.time()
 
+        # forward/backward pass (so)
```
That's lots of repeated code. Why not loop over ```["sp_", "s_o", "_po"]``` and use an if-elif-else for computing scores/loss?

vonVogelstein commented:
Adjusted variable names, added feature request (allow all combinations of query types) and adjusted all code to work on dictionaries over query types instead of individual variables per query type

rgemulla commented on kge/job/train.py:
```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
             # now create the results
-            sp_po_batch = torch.zeros([len(batch), 2], dtype=torch.long)
-            is_sp = torch.zeros([len(batch)], dtype=torch.long)
+            key_batch = torch.zeros([len(batch), 2], dtype=torch.long)
```
-> queries

rgemulla commented on kge/job/train.py:
```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
-
-                sp_po_batch[batch_index,] = keys[example_index]
+                prev_end = 0
+                for query_index, (query, end) in enumerate(self.query_end_index.items()):
```
Likewise, any order assumed?

vonVogelstein replied:
any

vonVogelstein replied:
any

rgemulla commented on kge/job/train.py:
```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
+                        keys = self.keys[query]
+                        offsets = self.offsets[query]
+                        values = self.values[query]
+                        if query_index == 0:
```
Woudln't it be more natural if query_index refers to the position of the element being predicted? (`0=_po, 1=s_o, 2=sp_`)?

rgemulla commented on kge/config-default.yaml:
```diff
@@ -220,6 +220,12 @@ KvsAll:
+  queries:
+    sp_to_o: True
+    so_to_p: False
+    po_to_s: True
```
Perhaps we should change the naming to make everything more readable. E.g.

query_type: e.g., sp_, (use this in the config)
target: e.g., o (the _ part of a query_type)
query: an actual query, e.g. (AlbertEinstein, born_In)
labels: the result(s) of query 
example: a query + its labels

rgemulla commented on kge/job/train.py:
```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
-            self.train_po_values,
-            self.train_po_offsets,
-        ) = kge.indexing.prepare_index(train_po)
+        self.queries = self.config.get("KvsAll.queries")
```
Should be query_type with the suggested naming

rgemulla commented on kge/job/train.py:
```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
-            - pairs (nx2 tensor, row = sp or po indexes),
-            - label coordinates (position of ones in a batch_size x num_entities tensor)
-            - is_sp (vector of size n, 1 if corresponding example_index is sp, 0 if po)
+            - pairs (nx2 tensor, row = sp, po or so indexes),
```
pairs -> queries

rgemulla commented on kge/job/train.py:
```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
-            sp_po_batch = torch.zeros([len(batch), 2], dtype=torch.long)
-            is_sp = torch.zeros([len(batch)], dtype=torch.long)
+            key_batch = torch.zeros([len(batch), 2], dtype=torch.long)
+            query_type = torch.zeros([len(batch)], dtype=torch.long)
```
query_types

rgemulla commented on kge/job/train.py:
```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
+                        offsets = self.offsets[query]
+                        values = self.values[query]
+                        if query_index == 0:
+                            query_col_1, query_col_2, value_col = S, P, O
```
target_col

rgemulla commented on kge/job/train.py:
```diff
@@ -640,50 +650,61 @@ def collate(batch):
+
+        query_labels = {}
+        for query, indexes in key_indexes.items():
+            if query == "so_to_p":
```
This should become query_type == "s_o"

vonVogelstein commented on kge/job/train.py:
```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
-                    num_ones -= self.train_po_offsets[example_index]
+                prev_end = 0
+                for query, end in self.query_end_index.items():
+                    if example_index < end:
```
relies on dict to be in some order (which is always the case in python)

rgemulla replied:
Is it? While I do think that this is the case in any likely implementation as long as the keyset is not changed, making such an assumption makes the code hard to read. The value of "end" depends on the order of the keys in the iteration.

More readable may be to use a map from query to (begin,end) or an ordered array of (end_index, query_type) pairs. An ordereddict would also be fine, but that still makes the code somewhat unaccessible.

rgemulla replied:
I had another look. The code sets `self.query_end_index[query] = prev_end + len(data[query])`. It implicitly assumes that `query_end_index` retains insertion order / is ordered the same way as `queries`. That's a hard-to-grasp assumption and it may not be valid.

vonVogelstein replied:
As far as I can see insertion into `self.query_end_index` is based on a loop over `queries` and `self.query_end_index` is a dict, therefore retaining insertion order according to the order in queries. Anyhow, I will adjust the code to be easier to understand

vonVogelstein commented:
As discussed used new naming conventions and worked with list instead of dict where required

rgemulla commented:
Thanks! I pushed some further revisions into your branch. Now let's reverse our roles: Please check if everything is OK and you agree with these changes.

vonVogelstein commented on kge/job/train.py:
```diff
@@ -530,37 +530,49 @@ def __init__(self, config, dataset, parent_job=None):
+            if enabled
+        ]
+
+        #' for each query type: list of queries
+        self.queries = {}
```
definition of query should be given for easier understanding

vonVogelstein commented on kge/job/train.py:
```diff
@@ -530,37 +530,49 @@ def __init__(self, config, dataset, parent_job=None):
+        #' for each query type: list of queries
+        self.queries = {}
+
+        #' for each query type: list of all labels (concatenated across queries)
+        self.labels = {}
```
definition of label should be given for easier understanding

vonVogelstein commented on kge/job/train.py:
```diff
@@ -530,37 +530,49 @@ def __init__(self, config, dataset, parent_job=None):
+        #' label_offsets[query_type][i]:label_offsets[query_type][i+1]
+        self.label_offsets = {}
+
+        #' for each query type (ordered as in self.query_types), index of last example
+        #' of that type in the list of all examples
+        self.query_end_index = []
```
Is it not the index of the last query of that type in the list of all queries (according to new definitions)?
Also, it is not the index of the last query but `1+ index of last query`, therefore the end in terms of pythonic slicing

vonVogelstein commented on kge/job/train.py:
```diff
@@ -619,85 +632,105 @@ def collate(batch):
+        label_coords_batch = batch["label_coords"].to(self.device)
+        query_type_indexes_batch = batch["query_type_indexes"]
+
+        examples_for_query_type = {}
+        for query_type_index, query_type in enumerate(self.query_types):
+            examples_for_query_type[query_type] = (
+                (query_type_indexes_batch == query_type_index)
+                .nonzero()
+                .to(self.device)
+                .view(-1)
+            )
```
should be `queries_for_query_type`

rgemulla commented:
Thanks! Any more comments on this?

vonVogelstein commented:
No, everything looks fine to me

rgemulla commented:
Thanks for your help!
